### PR TITLE
Pathfinder

### PR DESCRIPTION
**Pull request by miclegr**
ref #157

### CODE REVIEW

**Review by junpenglao — COMMENTED** [review]
Leaving the first round of comments, largely style wise. Need a bit more indepth review of the pathfinder implementation (vi/pathfinder.py)
There are a lot of places single letter variable names are used. Suggest changing to more informative name, for example `x` -> `state`, `x0` -> `initial_state`, `J` -> `param_dims`, `g` -> `grad`

File: blackjax/adaptation/pathfinder_adaptation.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        pathfinder_rng_key, sample_rng_key = jax.random.split(rng_key, 2)
+        pathfinder_state = pathfinder_init_fn(
+                pathfinder_rng_key,
+                logprob_fn,
+                initial_position)
+        new_initial_position = sample_from_state(
+                sample_rng_key,
+                pathfinder_state,
+                1)[0]
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/adaptation/pathfinder_adaptation.py]: I understand, not liking it too. I tried to implement your idea but I couldn't due to the need to support pytrees as positions. Basically in `sample_from_state` we ravel the pytree and then map the unravel function over that leading axis. If we drop the leading axis in `lbfgs_sample` then we would need to implement a different logic based on `M` being 1 or >1. That would do more harm then good I think.

This made me realize that that `[0]` is actually wrong and will raise an error if `pathfinder_state.x` is a pytree instead of an array. Hence I've change the code into calling `jax.map_pyteee`. I think it's a bit more clear now but still hackish

File: blackjax/kernels.py [hunk 2]
```diff
@@ -373,7 +376,7 @@ def kernel_fn(rng_key, state):
     schedule = adaptation.window_adaptation.schedule(num_steps)
     init, update, final = adaptation.window_adaptation.base(
         kernel_factory,
-        is_mass_matrix_diagonal,
+        logprob_fn,
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/kernels.py]: Ups, sorry for that :)

File: blackjax/kernels.py [hunk 3]
```diff
@@ -463,3 +466,163 @@ def step_fn(rng_key: PRNGKey, state):
...
+    """
+
+    init = staticmethod(vi.pathfinder.init)
+    kernel = staticmethod(vi.pathfinder.kernel)
+
+    def __new__(  # type: ignore[misc]
+        cls,
+        rng_key: PRNGKey,
+        logprob_fn: Callable,
+        M: int = 200,
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/kernels.py]: Yep, I agree with your observation to have more meaningful variable names. 
I originally set them to match the paper notation but now I changed user facing ones with more informative labels.
The not user facing variables are still matching the paper notation, for easier check of the formulas.

File: blackjax/kernels.py [hunk 4]
```diff
@@ -463,3 +466,163 @@ def step_fn(rng_key: PRNGKey, state):
...
+
+        def init_fn(position: PyTree):
+            return cls.init(
+                    rng_key,
+                    logprob_fn,
+                    position,
+                    M,
+                    False,
+                    **lbfgs_kwargs
+                    )
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/kernels.py]: Nope, forgot about that, did it now

File: blackjax/kernels.py [hunk 5]
```diff
@@ -463,3 +466,163 @@ def step_fn(rng_key: PRNGKey, state):
...
+        )
+        last_chain_state, last_warmup_state = last_state
+        history_state, history_info, history_da = warmup_chain
+
+        warmup_chain = (
+                history_state,
+                history_info,
+                last_warmup_state._replace(
+                    da_state=history_da
+                    )
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/kernels.py]: done

File: blackjax/vi/pathfinder.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+        maxcor: float = 10,
+        ftol: float = 2.220446049250313e-09,
+        gtol: float = 1e-05,
+        maxfun: int = None,
+        maxgrad: int = None,
+        maxls: int = 20,
+        ) -> Tuple[LBFGSResults, LBFGSHistory]:
+    """
+    Licensed under the Apache License, Version 2.0 (the "License");
+    TAKEN FROM https://github.com/google/jax/blob/main/jax/_src/scipy/optimize/_lbfgs.py
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/vi/pathfinder.py:237]: Ok, will open an issue in the Jax repo

File: blackjax/vi/pathfinder.py [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+
+
+def ELBO(logp, logq):
+    """
+    Calculates approximate ELBO using monte carlo.
+    It implements algorithm of figure 9 in:
+
+    Pathfinder: Parallel quasi-newton variational inference, Lu Zhang et al., arXiv:2108.03782
+    """
+    return (logp - logq).mean()
```
[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/vi/pathfinder.py]: Done

**Comment by miclegr** [comment]
> @miclegr Thank you for contributing, this looks like it will be a valuable contribution to the library. Could you please rebase `main` onto your branch? When you checked out to create your branch the CI tests were not properly configured and thus are not running.

sure, did the rebasing now. Please note that when it comes to tests there's the problem I've outlined in my comment in #157 :

> Unfortunately the L-BFGS optimizer is not working well when optimizing model's (negative) log-lilekihood while working in jax's default float32 mode, usually convergence fails. I've noticed that by simply dividing model's likelihood by number of observations (hence optimizing model's average likelihood) optimization converges. Unfortunately given blackjax design it's quite unnatural to ask for average likelihood. So in the end it's recommended to turn on double precision mode

> A consequence of this is that, for running pathfinder tests, double precision mode is needed. Since double precision mode needs to be set at jax initialization time (see [here](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision)), test suite should support some test in float32 mode and some in float64 mode (e.g. by running them in separate processes). It's feasible (see [here](https://stackoverflow.com/a/48310939) for example) but not implemented at the moment. Right know i've just set the float64 mode for that tests: https://github.com/miclegr/blackjax/blob/128ce00bd2b28e06f79d126c9d3c097b6378ccc8/tests/test_pathfinder.py#L2-L3 But this wont work when running the full suite of tests. Happy to spend some time to get the multi float mode tests in place, if that's the solution.

File: blackjax/adaptation/pathfinder_adaptation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (junpenglao) [reply blackjax/adaptation/pathfinder_adaptation.py]: I see. In that case could you add a comment explaining why we need indexing here?

[the remark this answers is not in the document]
Comment (miclegr) [reply blackjax/adaptation/pathfinder_adaptation.py]: Done

**Comment by miclegr** [comment]
Tentative feature request to have L-BFGS optimization path exposed directly in jax, without the code duplication of this pull request: google/jax#10243

**Comment by junpenglao** [comment]
Per discussion in https://github.com/google/jax/issues/10243, seems it is a better practice to use JAXopt - could you update the code and add a JAXopt dependency?

**Comment by miclegr** [comment]
Yep, I'm going to start working on that. Do you prefer JAXopt as a full requirement (i.e. as an entry in requirements.txt) or as an additional one (i.e. import within a try/except)?

**Comment by junpenglao** [comment]
let's make it a full requirement - I imagine we will use it in other optimization related inference method

**Comment by rlouf** [comment]
@miclegr is there anything we can do to help?

**Comment by miclegr** [comment]
Hi! I was finally able to complete the refactoring to make the code use `jaxopt` instead of `jax.scipy.optimize`. 
I've also took some time to make it work as much as possible with float32, by making ad ad-hoc adjustment to the stepsize of the LBFGS during training: https://github.com/miclegr/blackjax/blob/pathfinder/blackjax/vi/pathfinder.py#L290-L295

Now the tests run fine without the need to turn on the float64 mode, also the notebook :) So it should be mergeable.
Happy to do any remaing work

Michele

**Comment by junpenglao** [comment]
Great work @miclegr! I made some minor clean up - will merge once test pass.

**Comment by miclegr** [comment]
Thanks! and thanks for all the great work at pymc and tensorflow_probability 😉
